github_test.go

```diff
@@ -0,0 +1,370 @@
+package checkup
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/go-github/github"
+)
+
+var checkFileRegexp = regexp.MustCompile(`\A\d+-check.json\z`)
+
+func mustWriteJSON(w io.Writer, data interface{}) {
+	if err := json.NewEncoder(w).Encode(data); err != nil {
+		panic(err)
+	}
+}
+
+func base64Encoded(input []byte) string {
+	return base64.StdEncoding.EncodeToString(input)
+}
+
+func toJSON(data interface{}) []byte {
+	encoded, err := json.Marshal(data)
+	if err != nil {
+		panic(err)
+	}
+	return encoded
+}
+
+func sha(data []byte) string {
+	return fmt.Sprintf("%x", sha1.Sum(data))
+}
+
+func TestGitHub(t *testing.T) {
+	results := []Result{{Title: "Testing"}}
+	resultsBytes := []byte(`[{"title":"Testing"}]` + "\n")
+
+	// test server
+	mux := http.NewServeMux()
+	server := httptest.NewServer(mux)
+
+	// github client configured to use test server
+	client := github.NewClient(nil)
+	url, _ := url.Parse(server.URL)
+	client.BaseURL = url
+	client.UploadURL = url
+	defer server.Close()
+
+	// Our subject, our specimen.
+	specimen := GitHub{
+		RepositoryOwner: "o",
+		RepositoryName:  "r",
+		CommitterName:   "John Appleseed",
+		CommitterEmail:  "appleseed@example.org",
+		Branch:          "b",
+		Dir:             "subdir/",
+		client:          client,
+	}
+
+	// files := map[string]string{}
+	index := map[string]int64{
+		"1501523631505010894-check.json": time.Now().UnixNano(),
+		"1501525202306053005-check.json": time.Now().UnixNano(),
+	}
+	gitRepo := struct {
+		LastUpdated int64
+		Files       map[string]bool
+	}{
+		LastUpdated: time.Now().UnixNano(),
+		Files: map[string]bool{
+			"subdir/1501523631505010894-check.json": true,
+			"subdir/1501525202306053005-check.json": true,
+			"subdir/index.json":                     true,
+		},
+	}
+
+	serverSHAForRepo := sha(toJSON(gitRepo))
+
+	mux.HandleFunc("/repos/o/r/git/refs/heads/"+specimen.Branch, func(w http.ResponseWriter, r *http.Request) {
+		fmt.Printf("Processing %s %s\n", r.Method, r.URL.Path)
+		if got := r.Method; got != "GET" {
+			t.Errorf("Request method: %v, want %v", got, "GET")
+		}
+		mustWriteJSON(w, github.Reference{
+			Ref: github.String("refs/heads/" + specimen.Branch),
+			Object: &github.GitObject{
+				Type: github.String("commit"),
+				SHA:  github.String(serverSHAForRepo),
+			},
+		})
+	})
+
+	mux.HandleFunc("/repos/o/r/git/trees/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Printf("Processing %s %s\n", r.Method, r.URL.Path)
```

Comment:
Print statements can go away in the tests too, I think. Keeping the output uncluttered.